_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,503 @@
+---
+layout: post
+title: keywords: future workloads scheduling
+author: jguitana
+---
+
+One of the most relevant Scala primitives for concurrency is the `Future`. It represents a value which is either complete or failed, and is scheduled to run outside the program main flow, in some other thread.
+
+Future enables us to tackle the following concurrency problems:
+1. Let other components execute when one is waiting on IO.
+2. Make small programs communicate to build larger ones.
+3. Use all the CPU's in our machines.
+4. Enable human-computer interaction (GUIs, terminal).
+
+In this post we will look at three distinct kinds of workloads we can run in a Future: `cpu bound`, `blocking` and `asynchronous`, and gain some intuition about how they perform.
+
+These workloads generally fit in the problems described above as follows:
+1. blocking, async
+2. blocking, async
+3. cpu bound
+4. async
+
+Let's start by introducing each workload in more detail.
+
+#### Flavors of Future
+
+A `cpu bound` method will be using the cpu all the time. The following simulates such a method.
+
+```
+def cpuBound(v: Int): Int = {
+    val start = System.currentTimeMillis()
+    while ((System.currentTimeMillis() - start) < v) {}
+    println(Thread.currentThread())
+    v
+}
+```
+
+We can create a Future with a cpu bound workload that will be delegated to another thread. The OS will then schedule that threads workload to run in one of our cores.
```

Comment:
> The OS will then schedule that threads workload to run in one of our cores.

Seems a bit weird when reading, do you mean `The OS will then schedule which threads workload to run in one of our cores.`?

Reply:
Yeah will change it, it reads weird